core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -195,9 +208,30 @@ private synchronized void closeWithTragicEvent(Exception exception) throws IOExc
 
         operationCounter++;
 
+        assert assertNoSeqNumberConflict(seqNo, data);
+
         return new Translog.Location(generation, offset, data.length());
     }
 
+    private synchronized boolean assertNoSeqNumberConflict(long seqNo, BytesReference data) throws IOException {
+        if (seqNo == SequenceNumbersService.UNASSIGNED_SEQ_NO) {
+            // nothing to do
+        } else if (seenSequenceNumbers.containsKey(seqNo)) {
+            final Tuple<BytesReference, Exception> previous = seenSequenceNumbers.get(seqNo);
```

Comment:
I did it this way so it can easily integrate into a caused by clause and have a proper message. Do you have a better suggestion  there?